Limit Apiserver Proxy Redirects

riaankleinhans opened the pull request:
*What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**

**Which issue(s) this PR fixes:**
Fixes #95129

**Testgrid Link:**

**Special notes for your reviewer:**
```
NONE

```

**Does this PR introduce a user-facing change?:**
```
Yes
```

**Release note:**
```release-note
kube-apiserver: requests to node, service, and pod `/proxy` subresources with no additional URL path now only automatically redirect GET and HEAD requests.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/networking

fedebongio commented:
/assign @deads2k

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:
```diff
@@ -207,7 +208,7 @@ func (h *UpgradeAwareHandler) ServeHTTP(w http.ResponseWriter, req *http.Request
 	// This is essentially a hack for http://issue.k8s.io/4958.
 	// Note: Keep this code after tryUpgrade to not break that flow.
-	if len(loc.Path) == 0 {
+	if len(loc.Path) == 0 && (method == http.MethodGet || method == http.MethodHead) {
```
limiting the redirect to GET/HEAD requests looks good to me.

by continuing with the normal proxy path for other methods (e.g. PUT/POST), what is the HTTP request this change would end up passing to the underlying pod if `len(loc.Path)` was 0 here?

heyste replied:
From my tests the other methods will pass through to the pod unaffected. Using `kubectl proxy`, `porter` and `curl` I can get 200 OK response back from the pod for the other http methods.
```
$ curl -I -X POST http://localhost:9999/api/v1/namespaces/default/pods/agnhost-7bb97c7fb7-6k9dq/proxy
HTTP/1.1 200 OK
Cache-Control: no-cache, private
Content-Length: 3
Content-Type: text/plain; charset=utf-8
Date: Wed, 30 Sep 2020 18:43:36 GMT
```

heyste replied:
Using this PR with an updated [e2e test](https://github.com/apisnoop/ticket-writing/blob/NamespacedPodProxy/e2e/proxy.go#L266-L341) for NamespacedPodProxy endpoints gives
```
STEP: Creating a Pod
Oct  1 01:17:33.579: INFO: pod created
Oct  1 01:17:33.591: INFO: Pod Quantity: 1 Status: Pending
Oct  1 01:17:34.596: INFO: Pod Quantity: 1 Status: Pending
Oct  1 01:17:35.597: INFO: Pod Quantity: 1 Status: Pending
Oct  1 01:17:36.596: INFO: Pod Status: Running
Oct  1 01:17:36.596: INFO: Starting http.Client for https://127.0.0.1:41813/api/v1/namespaces/proxy-9664/pods/proxy-pod-target/proxy
Oct  1 01:17:36.600: INFO: http.Client request:GET StatusCode:301
Oct  1 01:17:36.603: INFO: http.Client request:HEAD StatusCode:301
Oct  1 01:17:36.608: INFO: http.Client request:DELETE StatusCode:200
Oct  1 01:17:36.612: INFO: http.Client request:OPTIONS StatusCode:200
Oct  1 01:17:36.616: INFO: http.Client request:PATCH StatusCode:200
Oct  1 01:17:36.621: INFO: http.Client request:POST StatusCode:200
Oct  1 01:17:36.624: INFO: http.Client request:PUT StatusCode:200
```

liggitt replied:
> what is the HTTP request this change would end up passing to the underlying pod if len(loc.Path) was 0 here?

I wanted to make sure the request sent to the underlying pod was a valid HTTP request (especially that a non-empty request-uri was sent to the backing pod)

https://www.w3.org/Protocols/HTTP/1.1/rfc2616bis/draft-lafon-rfc2616bis-03.html#request-uri

> Note that the absolute path cannot be empty; if none is present in the original URI, it must be given as "/" (the server root).

heyste replied:
- Updated PR so 'absolute path cannot be empty' now.
- Checked how the pod handles the request & payload sent back to the client.
- Tested with `kubectl proxy`, `curl`, `porter` and `tcpdump` 
```
    10.244.0.1.42256 > 10.244.0.5.80: Flags [P.], cksum 0x16e4 (incorrect -> 0x7749), seq 211:419, ack 120, win 502, options [nop,nop,TS val 1754840149 ecr 1101893751], length 208: HTTP, length: 208
        DELETE / HTTP/1.1
        Host: localhost:9999
        User-Agent: via-proxy
        Accept: */*
        Accept-Encoding: gzip
        X-Forwarded-For: 127.0.0.1, 172.18.0.1
        X-Forwarded-Uri: /api/v1/namespaces/default/pods/agnhost2/proxy/

02:58:21.755520 IP (tos 0x0, ttl 64, id 11616, offset 0, flags [DF], proto TCP (6), length 52)
    10.244.0.5.80 > 10.244.0.1.42256: Flags [.], cksum 0x1614 (incorrect -> 0x25f4), ack 419, win 507, options [nop,nop,TS val 1101914757 ecr 1754840149], length 0
02:58:21.755774 IP (tos 0x0, ttl 64, id 11617, offset 0, flags [DF], proto TCP (6), length 171)
    10.244.0.5.80 > 10.244.0.1.42256: Flags [P.], cksum 0x168b (incorrect -> 0x5f0d), seq 120:239, ack 419, win 507, options [nop,nop,TS val 1101914758 ecr 1754840149], length 119: HTTP, length: 119
        HTTP/1.1 200 OK
        Date: Tue, 06 Oct 2020 02:58:21 GMT
        Content-Length: 3
        Content-Type: text/plain; charset=utf-8

        foo[!http]
```

lavalamp replied:
Why would we limit to GET & HEAD? Why couldn't people be using POST PUT etc? Isn't this a compatibility breakage?

liggitt replied:
The request body associated with a PUT/POST/PATCH/etc request doesn't get sent to the backend, and the clients I know of drop their PUT/POST/PATCH content in response to a 301 request, and switch to making a GET request to the new location (including [golang](https://github.com/golang/go/blob/0d65c272c9e494cbb604f2bee99d434b8cde46ff/src/net/http/client.go#L505-L516) and [browsers](https://stackoverflow.com/a/13628908/54696), which is what this redirect was originally added for)

That makes returning a 301 redirect for anything other than GET/HEAD lose data and not be usable in the scenarios I can see

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:
```diff
@@ -212,9 +213,15 @@ func (h *UpgradeAwareHandler) ServeHTTP(w http.ResponseWriter, req *http.Request
+			return
+		default:
+			path := req.URL.Path
+			req.URL.Path = path + "/" + queryPart
```
this is doing a few things that don't seem quite right:

1. it is adding duplicating query data into the path (I think a request with query parameters would get an incorrect path, need to add a unit test to exercise this)
2. it is modifying `req.URL.Path` but not `loc.Path` (which means the `newReq.URL = &loc` code path below is still wrong... need a unit test to exercise that path as well)

My question in https://github.com/kubernetes/kubernetes/pull/95128#discussion_r498275061 was more of a question about behavior ... I'm not actually sure we need to append a `/` ... the go http client might default to sending `/` if there is not path present... if this default case is simply dropped, does the pod still get a `$METHOD / HTTP/1.1` call?

Riaankl commented:
/remove-lifecycle stale

Riaankl commented:
/assign @smarterclayton

Riaankl commented:
This PR is needed to move #94786 forward.

liggitt commented:
the comment at https://github.com/kubernetes/kubernetes/pull/95128/files#r500268226 still looks outstanding

Riaankl commented:
/remove-lifecycle stale
/cc @heyste

Riaankl commented:
/remove-lifecycle stale
/cc @heyste

heyste commented:
/test pull-kubernetes-node-e2e-containerd
unrelated flake `E2eNode Suite: [sig-storage] HostPath should support r/w [NodeConformance]`

heyste commented:
After retesting the original fix with kind, `kubectl proxy` and a custom porter, the following output shows that no empty request is sent to the pod (i.e a missing slash).

```
$ curl -XPOST http://127.0.0.1:11111/api/v1/namespaces/default/pods/debug-proxy/proxy && echo
{"Method":"POST","RequestURI":"/","Body":"foo"}
 
$ curl -XPOST http://127.0.0.1:11111/api/v1/namespaces/default/pods/debug-proxy/proxy?id=123 && echo
{"Method":"POST","RequestURI":"/?id=123","Body":"foo"}

$ curl -XDELETE http://127.0.0.1:11111/api/v1/namespaces/default/pods/debug-proxy/proxy && echo
{"Method":"DELETE","RequestURI":"/","Body":"foo"}

$ curl -XDELETE http://127.0.0.1:11111/api/v1/namespaces/default/pods/debug-proxy/proxy?id=123 && echo
{"Method":"DELETE","RequestURI":"/?id=123","Body":"foo"}
```
Checking APISnoop results
```
#+begin_src sql-mode :eval never-export :exports both :session none
select distinct  endpoint, right(useragent,83) AS useragent
from testing.audit_event
where endpoint ilike '%proxy%'
order by endpoint
limit 10;
#+end_src

#+RESULTS:
#+begin_SRC example
               endpoint                |  useragent
---------------------------------------+-------------
 connectCoreV1DeleteNamespacedPodProxy | curl/7.68.0
 connectCoreV1PostNamespacedPodProxy   | curl/7.68.0
(2 rows)

#+end_SRC
```
@liggitt Can you please confirm that the logs above do  [address your outstanding query](https://github.com/kubernetes/kubernetes/pull/95128#discussion_r498275061) and the PR can now be lgtm & approved? TIA

liggitt commented:
thanks for checking the impact on backend behavior, the change lgtm

unit test coverage of client side behavior (redirect received for get/head, no redirect received for put/post/delete/etc) and backend behavior (request with `/` received for all assuming the client followed redirects) would ensure the behavior you manually verified doesn't regress in the future

smarterclayton commented:
If jordan's question is answered to his satisfaction this is approved and also lgtm, will tag as such unless he comments otherwise in next day.

/approve

liggitt commented:
the change lgtm but without test coverage nothing would catch a regression

heyste commented:
@liggitt  For the new unit test to check the status code (either 301/200) for various methods I'm having an issue getting passed the following code block

https://github.com/kubernetes/kubernetes/blob/92ddd4dcd62810d4fed67abc54dd114e21776161/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L210-L212

Using the current `TestServeHTTP` scenario with a `requestPath:  ""` it will get replaced with a `/` which means it will then skips the next check

https://github.com/kubernetes/kubernetes/blob/92ddd4dcd62810d4fed67abc54dd114e21776161/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L218

Question: Based on the output below does that get processed somewhere inside of [pkg/registry/core/pod/rest/subresources.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/rest/subresources.go) or have I missed something in the process of getting from `curl` to the `pod` ?

```
curl -I http://localhost:12345/api/v1/namespaces/default/pods/agnhost2-5b47dc9979-tfdj5/proxy
HTTP/1.1 301 Moved Permanently
Audit-Id: 42beb19b-50e8-4491-a7de-9ed5c9d6a842
Cache-Control: no-cache, private
Date: Tue, 21 Sep 2021 23:33:39 GMT
Location: /api/v1/namespaces/default/pods/agnhost2-5b47dc9979-tfdj5/proxy/
```

TIA for any tips you can provide.

liggitt commented:
> Using the current TestServeHTTP scenario with a requestPath: "" it will get replaced with a /

You can't send a completely empty path to a server. If the test server roots the proxy handler at `/`, I don't think there's a way to entirely omit a path in a request to it.

In real use, the proxy handler is rooted at `/api/v1/namespaces/$ns/pods/$pod/proxy`, so a request to ``/api/v1/namespaces/$ns/pods/$pod/proxy` (no trailing slash) would be handled by the proxy handler, would have the `/api/v1/namespaces/$ns/pods/$pod/proxy` prefix removed, and would end up with an empty path with no trailing slash it was expected to proxy through.

heyste committed:
Redirect proxy requests for only GET & HEAD methods

- Extract the current redirect code into a function
  (proxyRedirectsforRootPath) and redirect for only
  http.MethodGet or http.MethodHead
- Create a unit test that confirms that only GET &
  HEAD methods are redirected

heyste commented:
/test pull-kubernetes-node-e2e-containerd
unrelated flake
```
W1006 03:17:04.446] Could not detect Kubernetes master node.  Make sure you've launched a cluster with 'kube-up.sh'
I1006 03:17:04.546] Master not detected. Is the cluster up? 
```

liggitt commented:
/lgtm
/approve

heyste commented:
/test pull-kubernetes-e2e-gce-100-performance
unrelated flake `e2e.go:Up`
```
scp: /var/log/cluster-autoscaler.log*: No such file or directory
scp: /var/log/fluentd.log*: No such file or directory
scp: /var/log/kubelet.cov*: No such file or directory
scp: /var/log/cl2-**: No such file or directory
scp: /var/log/startupscript.log*: No such file or directory
ERROR: (gcloud.compute.scp) [/usr/bin/scp] exited with return code [1]. 
```
